modules/contracts/apps/challenge/src/handlers/execute.rs

```diff
@@ -0,0 +1,385 @@
+            cancel_challenge(deps, info, app, challenge_id)
+        }
+        ChallengeExecuteMsg::AddFriendForChallenge {
+            challenge_id,
+            friend_name,
+            friend_address,
+        } => add_friend_for_challenge(
+            deps,
+            info,
+            &app,
+            challenge_id,
+            &friend_name,
+            &friend_address,
+        ),
+        ChallengeExecuteMsg::RemoveFriendForChallenge {
+            challenge_id,
+            friend_address,
+        } => remove_friend_from_challenge(deps, info, &app, challenge_id, friend_address),
+        ChallengeExecuteMsg::AddFriendsForChallenge {
+            challenge_id,
+            friends,
+        } => add_friends_for_challenge(deps, info, &app, challenge_id, friends),
+        ChallengeExecuteMsg::DailyCheckIn {
+            challenge_id,
+            metadata,
+        } => daily_check_in(deps, env, info, &app, challenge_id, metadata),
+        ChallengeExecuteMsg::CastVote { vote, challenge_id } => {
+            cast_vote(deps, env, info, &app, vote, &challenge_id)
+        }
+        ChallengeExecuteMsg::CountVotes { challenge_id } => {
+            count_votes(deps, info, env, &app, challenge_id)
+        }
+    }
+}
+
+/// Create new challenge
+fn create_challenge(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    app: ChallengeApp,
+    challenge: ChallengeEntry,
+) -> AppResult {
+    // Only the admin should be able to create a challenge.
+    app.admin.assert_admin(deps.as_ref(), &info.sender)?;
+
+    // Generate the challenge id
+    let id = NEXT_ID.update(deps.storage, |id| AppResult::Ok(id + 1))?;
+    let challenge_id = format!("challenge_{id}");
+
+    CHALLENGE_LIST.save(deps.storage, challenge_id.clone(), &challenge)?;
+
+    Ok(app.tag_response(
+        Response::new().add_attribute("challenge_id", challenge_id),
+        "create_challenge",
+    ))
+}
+
+/// Update an existing challenge  
+fn update_challenge(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    app: ChallengeApp,
+    challenge_id: String,
+    challenge: ChallengeEntry,
+) -> AppResult {
+    app.admin.assert_admin(deps.as_ref(), &info.sender)?;
+
+    // will return an error if the challenge doesn't exist
+    CHALLENGE_LIST.load(deps.storage, challenge_id.clone())?;
+    CHALLENGE_LIST.save(deps.storage, challenge_id.clone(), &challenge)?;
+
+    Ok(app.tag_response(
+        Response::new().add_attribute("challenge_id", challenge_id),
+        "update_challenge",
+    ))
+}
+
+fn cancel_challenge(
+    deps: DepsMut,
+    info: MessageInfo,
+    app: ChallengeApp,
+    challenge_id: String,
+) -> AppResult {
+    app.admin.assert_admin(deps.as_ref(), &info.sender)?;
+
+    CHALLENGE_LIST.load(deps.storage, challenge_id.clone())?;
+    CHALLENGE_LIST.remove(deps.storage, challenge_id.clone());
+
+    Ok(Response::new().add_attribute("action", "cancel_challenge"))
```

Comment:
Could be nice to have this return the ID as well in the attributes.

Reply:
Is there any reason to load it by the way? `remove()` won't do anything if there was no value by this key

Reply:
> Could be nice to have this return the ID as well in the attributes.

Hmm, I would disagree with that, since `challenge_id` will already be logged in tx, no need to repeat it

Reply:
I guess, not sure if indexers can easily make use of those values tho.